apps/web/middleware.ts

```diff
@@ -100,9 +136,46 @@ const middleware = async (req: NextRequest): Promise<NextResponse<unknown>> => {
     },
   });
 
+  const firstDomainInPathname = `/${url.pathname.split("/")[1]}`;
+
+  if (!NON_BOOKING_ROUTES_IN_APP_ROUTER.includes(firstDomainInPathname)) {
+    // either /team/** pages, or /org/** pages, or /[user]/** pages
+    const csrfToken = requestHeaders.get("x-csrf-token") || "missing";
+    res.cookies.set("x-csrf-token", `${csrfToken}`, {
+      httpOnly: true,
+      path: "/",
+      sameSite: "lax",
+      secure: process.env.NODE_ENV === "production",
+    });
+
+    const csrfResponse = await handleCsrfProtect(req, res);
+    if (csrfResponse) return csrfResponse;
+  }
+
   return responseWithHeaders({ url, res, req });
 };
 
+async function handleCsrfProtect(req: NextRequest, res: NextResponse) {
+  const url = req.nextUrl;
+  // Skip CSRF protection for trpc requests for now. Prevents E2E tests from failing.
+  // Most trcp endpoints are authenticated.
+  if (url.pathname.startsWith("/api/trpc/")) return;
+  // NextAuth handles CSRF protection for itself
+  if (url.pathname.startsWith("/api/auth/")) return;
+  if (url.pathname.startsWith("/_next")) return;
+  if (req.method === "GET") return;
```

Comment:
added these 2 filtering as well @zomars

Reply:
If we skip GET then we will get a "missing" token. Headers won't be set so there's no token to save to the cookie.